### ISSUE TITLE
error using with 2.0.0-alpha.1

### ISSUE THREAD

**Issue by xmlking** [alexmingoia/koa-router#199]
I am testing `koa` version `2.0.0-alpha.1` with `koa-router` and got `RangeError: Maximum call stack size exceeded`

**Failing code**
```js
import Koa from 'koa';
import Router from 'koa-router';


let rootRouter = new Router({
  prefix: '/api'
});

const app = new Koa();

rootRouter.use('/test', async (ctx, next) => {
  const start = new Date;
  await next();
  const ms = new Date - start;
  console.log(`${ctx.method} ${ctx.url} - ${ms}`);
});

rootRouter.use(function (ctx){
  ctx.body = 'Hello World';
});

app.use(rootRouter.routes());
app.listen(3000);
```

**error**
```
$ DEBUG=koa-router npm start

> demo@1.1.0 start /Developer/Work/node/demo
> babel-node test.js

  koa-router defined route  /api/test +0ms
  koa-router defined route  /api(.*) +2ms
  koa-router undefined undefined +9s
  koa-router test /api/test /^\/api\/test(?:\/(?=$))?(?=\/|$)/i +0ms
  koa-router test /api(.*) /^\/api(.*)(?:\/(?=$))?(?=\/|$)/i +0ms

  RangeError: Maximum call stack size exceeded
      at objectToPromise (/Developer/Work/node/demo/node_modules/co/index.js:167:17)
      at toPromise (/Developer/Work/node/demo/node_modules/co/index.js:121:45)
      at objectToPromise (/Developer/Work/node/demo/node_modules/co/index.js:172:29)
      at toPromise (/Developer/Work/node/demo/node_modules/co/index.js:121:45)
      at objectToPromise (/Developer/Work/node/demo/node_modules/co/index.js:172:29)
      at toPromise (/Developer/Work/node/demo/node_modules/co/index.js:121:45)
      at objectToPromise (/Developer/Work/node/demo/node_modules/co/index.js:172:29)
      at toPromise (/Developer/Work/node/demo/node_modules/co/index.js:121:45)
      at objectToPromise (/Developer/Work/node/demo/node_modules/co/index.js:172:29)
      at toPromise (/Developer/Work/node/demo/node_modules/co/index.js:121:45)

```

**Koa 2 working sample**
```js
const Koa = require('koa');
const app = new Koa();

// logger

app.use(async (ctx, next) => {
  const start = new Date;
  await next();
  const ms = new Date - start;
  console.log(`${ctx.method} ${ctx.url} - ${ms}`);
});

// response

app.use(function (ctx){
  ctx.body = 'Hello World';
});

app.listen(3000);
```

**Comment by felixfbecker** [comment]
I have the same issue. Apparently, koa-router doesn't support koa2 yet. Maybe this issue should be renamed into koa2 support? I would like to use koa2 but I can't without a good router.
Three things changed in koa2:
 - context is passed as arguments instead of `this`
 - `next()` must be called and returns a promise
 - generators are no longer supported, use promises or co.wrap
Are there plans to support koa2?

**Comment by menems** [comment]
Yes had the same problem, I wrote another router for now [koa-66](https://github.com/menems/koa-66)

**Comment by felixfbecker** [comment]
Is it API compatible? I saw the mount function which doesn't seem to be in koa-router. Also, is this basically a fork of koa-router or did you start from scratch?

**Comment by menems** [comment]
Almost compatible, not all features are present, like params, redirect and other things like that, and I decided to add a mount function instead of 'use', I did it from scratch... but i pick a lot from koa-router.
I didn't fork it because, i think algo of routes() methods it's difficult to change without broke the old compatibility, and think the use of a layer prototype is useless, but it's just a personal thought

**Comment by felixfbecker** [comment]
I need params, 405 and all that :/
Nice name though.

**Comment by menems** [comment]
405 is already present, yes maybe add router.param function on it, I thought

**Comment by alexmingoia** [comment]
koa-router@6 supports koa@2. koa 2 support was implemented in #202
